File: mf2/go/locale_direction.go
package mf2

import (
	"regexp"
	"strings"
)

var directionLanguage = regexp.MustCompile(`^[a-z]{2,8}$`)
var directionSubtag = regexp.MustCompile(`^[A-Za-z0-9]{2,8}$`)
var directionScript = regexp.MustCompile(`^[A-Za-z]{4}$`)
var directionRegion = regexp.MustCompile(`^([A-Za-z]{2}|[0-9]{3})$`)

func localeIsLtr(locale string) bool {
	subtags := strings.Split(strings.ReplaceAll(locale, "_", "-"), "-")
	language := strings.ToLower(subtags[0])
	if !directionLanguage.MatchString(language) {
		return false
	}
	script, region := "", ""
	for _, subtag := range subtags[1:] {
		if len(subtag) == 1 {
			break
		}
		if !directionSubtag.MatchString(subtag) {
			return false
		}
		if script == "" && directionScript.MatchString(subtag) {
			script = strings.ToUpper(subtag[:1]) + strings.ToLower(subtag[1:])
		} else if region == "" && directionRegion.MatchString(subtag) {
			region = strings.ToUpper(subtag)
		}
	}
	contains := func(table, value string) bool { return strings.Contains(table, " "+value+" ") }
	if script != "" {
		return contains(localeDirectionLTR_SCRIPTS, script)
	}
	if region != "" {
		key := language + "-" + region
		if contains(localeDirectionRTL_REGION_OVERRIDES, key) {
			return false
		}
		if contains(localeDirectionLTR_REGION_OVERRIDES, key) {
			return true
		}
	}
	return !(language == "und" && region == "") && contains(localeDirectionLTR_LANGUAGES, language)
}

File: mf2/php/src/LocaleDirection.php
<?php
declare(strict_types=1);
namespace Mojito\MessageFormat2\Internal;

function locale_is_ltr(string $locale): bool
{
    $subtags = explode('-', str_replace('_', '-', $locale));
    $language = strtolower(array_shift($subtags));
    if (!preg_match('/^[a-z]{2,8}$/D', $language)) return false;
    $script = $region = null;
    foreach ($subtags as $subtag) {
        if (strlen($subtag) === 1) break;
        if (!preg_match('/^[A-Za-z0-9]{2,8}$/D', $subtag)) return false;
        if ($script === null && preg_match('/^[A-Za-z]{4}$/D', $subtag)) {
            $script = ucfirst(strtolower($subtag));
        } elseif ($region === null && preg_match('/^(?:[A-Za-z]{2}|[0-9]{3})$/D', $subtag)) {
            $region = strtoupper($subtag);
        }
    }
    $contains = static fn(string $table, string $value): bool => str_contains($table, " $value ");
    if ($script !== null) return $contains(LocaleDirectionData::LTR_SCRIPTS, $script);
    if ($region !== null) {
        $key = "$language-$region";
        if ($contains(LocaleDirectionData::RTL_REGION_OVERRIDES, $key)) return false;
        if ($contains(LocaleDirectionData::LTR_REGION_OVERRIDES, $key)) return true;
    }
    return !($language === 'und' && $region === null) && $contains(LocaleDirectionData::LTR_LANGUAGES, $language);
}
